desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,1349 @@
+        b = 1 / jnp.atleast_1d(pitch)
+        for val in b:
+            add(ax.axhline(val, color="tab:purple", alpha=0.25, label=r"$1 / \lambda$"))
+        bp1, bp2 = jnp.atleast_2d(bp1, bp2)
+        for i in range(bp1.shape[0]):
+            bp1_i, bp2_i = map(_filter_not_nan, (bp1[i], bp2[i]))
+            add(
+                ax.scatter(
+                    bp1_i,
+                    jnp.full_like(bp1_i, b[i]),
+                    marker="v",
+                    color="tab:red",
+                    label="bp1",
+                )
+            )
+            add(
+                ax.scatter(
+                    bp2_i,
+                    jnp.full_like(bp2_i, b[i]),
+                    marker="^",
+                    color="tab:green",
+                    label="bp2",
+                )
+            )
+
+    ax.set_xlabel(r"Field line $\zeta$")
+    ax.set_ylabel(r"$\vert B \vert \sim 1 / \lambda$")
+    ax.legend(legend.values(), legend.keys())
+    if id is not None:
+        title = f"{title}. id = {id}."
+    ax.set_title(title)
+    if show:
+        plt.tight_layout()
+        plt.show()
+        plt.close()
+    return fig, ax
+
+
+def affine_bijection(x, a, b):
+    """[−1, 1] ∋ x ↦ y ∈ [a, b]."""
+    y = (x + 1) / 2 * (b - a) + a
+    return y
+
+
+def grad_affine_bijection(a, b):
+    """Gradient of affine bijection."""
+    dy_dx = (b - a) / 2
+    return dy_dx
+
+
+def automorphism_arcsin(x):
+    """[-1, 1] ∋ x ↦ y ∈ [−1, 1].
+
+    The gradient of the arcsin automorphism introduces a singularity that augments
+    the singularity in the bounce integral. Therefore, the quadrature scheme
+    used to evaluate the integral must work well on functions with large
+    derivative near the boundary.
+
+    The arcsin automorphism pulls points in [−1, 1] away from the boundary.
+    This can reduce floating point error if paired with a quadrature
+    scheme that is aggressive with placing nodes near endpoints, such as
+    Tanh-Sinh quadrature.
+
+    Parameters
+    ----------
+    x : Array
+        Points to transform.
+
+    Returns
+    -------
+    y : Array
+        Transformed points.
+
+    """
+    y = 2 * jnp.arcsin(x) / jnp.pi
+    return y
+
+
+def grad_automorphism_arcsin(x):
+    """Gradient of arcsin automorphism."""
+    dy_dx = 2 / (jnp.sqrt(1 - x**2) * jnp.pi)
+    return dy_dx
+
+
+grad_automorphism_arcsin.__doc__ += "\n" + automorphism_arcsin.__doc__
+
+
+def automorphism_sin(x, s=0, m=10):
+    """[-1, 1] ∋ x ↦ y ∈ [−1, 1].
+
+    The gradient of the sin automorphism is Lipschitz.
+    When this automorphism is used as the change of variable map for the bounce
+    integral, the Lipschitzness prevents generation of new singularities.
+    Furthermore, its derivative vanishes to zero slowly near the boundary,
+    which will suppress the large derivatives near the boundary of singular
+    integrals.
+
+    Therefore, this automorphism pulls the mass of the bounce integral away
```

Comment:
This is a strange docstring. You are adding technical terms and then terms like "mass" of the bounce integral.

Reply:
I'm not partial to this description, but I also don't quite understand the comment.

Most users won't change the default. If they want to define their own automorphism $a$, a description like this is useful for them to understand what's going on and how it pairs with quadrature rules.
For example, a tanh-sinh quadrature paired with a singularity removing transformation will perform poorly. Also it's not sufficient for $||\nabla a||$ to go to zero at the singularity; it must do this slowly, that is $||\nabla a|| \approx 0$ in a small region around the singularity. I'm sure a user could come to this conclusion by themselves, but it's helpful to have it documented.

- Mass of an integral is a formal thing. I've seen it used in probability (i.e. probability mass) and analysis (stein and sharkarchi fourier analysis). I don't think phrases like "the mass of the integral is concentrated at the singularity of the integrand" are uncommon.
- automorphism: Alternatives are:
    - bijective, differentiable map from [-1, 1] to [-1, 1] whose inverse is also differentiable in the interior of [-1, 1].
    - change of variables from [-1, 1] to [-1, 1]. This is simpler but I could see someone using an even transformation, which is not bijective, and will give incorrect results.
- lipshitzness of the gradient $||\nabla a ||\leq L$ for some constant $L$ is the requirement to not introduce new singularities